### PR TITLE
Newton raphson

### PR DESCRIPTION
**Pull request by TTitscher**
Here is an idea for a generic Newton-Raphson algorithm as a free function. The implementation relies on _a lot_ of template arguments for defining
- residual function
- derivative function
- norm function
- tolerance value
- optional info method
all collected in NuTo::NewtonRaphson::Problem

Please have a look at https://github.com/nutofem/nuto/blob/newton_raphson/test/math/NewtonRaphson.cpp to see its application to both a scalar valued and a vector valued problem.

Is this tooo cryptic all in all?

### CODE REVIEW

File: applications/benchmark/CMakeLists.txt [hunk 1]
```diff
@@ -40,3 +40,6 @@ if(ENABLE_GEOMETRYCONCRETE)
     target_link_libraries(DenseSpherePacking NuToMath NuToBase NuToGeometryConcrete ${Boost_LIBRARIES})
     add_test(benchmark::DenseSpherePacking ${CMAKE_CURRENT_BUILD_DIR}/DenseSpherePacking)
 endif()
+
+add_executable(NewtonRaphsonBenchmark NewtonRaphsonBenchmark.cpp)
+target_link_libraries(NewtonRaphsonBenchmark -lprofiler)
```
Comment (Psirus) [inline applications/benchmark/CMakeLists.txt]: I don't think we've agreed on making the google profile library a dependency. And if we had, this would not be the "proper way" to link to it.

File: src/math/Average.h [hunk 2]
```diff
@@ -0,0 +1,17 @@
+#pragma once
+namespace NuTo
+{
+
+//! @brief calculates the cumulative moving average
+struct Average
```
Comment (Psirus) [inline src/math/Average.h:6]: What does this have to do with the Newton-Raphson? Where do we use it?

File: src/math/LineSearch.h [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+    //! @brief actual line search implementation
+    //! @remark several return arguments are used to help inlining everything. A lot of time went into actual
+    //! benchmarking. Dunno why, but the a previous version that avoids the return arguments and returns various values
+    //! in a std::tuple was significantly slower (10%) in a benchmark for a scalar function
+    //! @param problem ... class that implements NormFunction, ResidualFunction and mTolerance
+    //! @param r ... residual, return argument, see remark
+    //! @param x ... value of the argument x, return argument, see remark
+    //! @param dx ... dx from the solver
+    template <typename TProblem, typename TX>
+    bool operator()(TProblem&& problem, TX* r, TX* x, TX dx) const
```
Comment (Psirus) [inline src/math/LineSearch.h:39]: I'm guessing there is a reason why these are pointers, but I can't guess what it is. Can you help me?

File: src/math/NewtonRaphson.h [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+    TDR DerivativeFunction;
+    TNorm NormFunction;
+    TTol mTolerance;
+    TInfo InfoFunction;
+};
+
+//! @brief defines the problem, basically just to enable automatic template deduction. If you
+//! create a Problem directly, you'll have to specify each template parameter. This methods avoids it.
+template <typename TR, typename TDR, typename TNorm, typename TTol, typename TInfo = VoidInfo>
+auto DefineProblem(TR residual, TDR derivative, TNorm norm, TTol tolerance, TInfo info = VoidInfo())
```
Comment (Psirus) [inline src/math/NewtonRaphson.h:49]: Cool pattern, haven't seen that before.

File: src/math/NewtonRaphson.h [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+//! @brief solves the Problem using the newton raphson iteration with linesearch
+//! @param problem type of the nonlinear problem
+//! @param x0 of the initial value for the iteration
+//! @param solver solver that provides a TX = solver.Solve(TNonlinearProblem::DR, TNonlinearProblem::R)
+//! @param maxIterations default = 20
+//! @param lineSearch line search algorithm, default = NoLineSearch, alternatively use NuTo::LineSearch()
+//! @param numIterations optionally returns the number of iterations required 
+template <typename TNonlinearProblem, typename TX, typename TSolver, typename TLineSearchAlgorithm = NoLineSearch>
+auto Solve(TNonlinearProblem&& problem, TX&& x0, TSolver&& solver, int maxIterations = 20,
+            TLineSearchAlgorithm&& lineSearch = NoLineSearch(), int* numIterations = nullptr)
```
Comment (Psirus) [inline src/math/NewtonRaphson.h:63]: I don't think `numIterations` are actually set anywhere.

**Review by Psirus — APPROVED** [review]
Good idea, well implemented, thoroughly tested. What's not to like? How likely do you think it is that we'll replace the alternate implementations strewn about?

File: src/math/NewtonRaphson.h (hunk 5 above)
Comment (TTitscher) [reply src/math/NewtonRaphson.h:63]: Oopsy daisy, i'll fix that.

File: src/math/NewtonRaphson.h (hunk 4 above)
Comment (TTitscher) [reply src/math/NewtonRaphson.h:49]: This _work around_ will be gone in c++17 with [class template deduction](http://en.cppreference.com/w/cpp/language/class_template_deduction).

File: src/math/LineSearch.h (hunk 3 above)
Comment (TTitscher) [reply src/math/LineSearch.h:39]: I like the syntax that is required by the caller. You'll have to call this method with
~~~cpp
linesearch(problem, &r, &x, dx);
~~~
indicating that `r` and `x` are return arguments. This would be hidden otherwise.

**Review by pmueller2 — APPROVED** [review]
Nothing to add, except a little question: If I have to use this to find the complex roots of a function (using std::complex) there are only minor additions to make right?

File: src/math/Average.h (hunk 2 above)
Comment (TTitscher) [reply src/math/Average.h:6]: Dunno. I wrote it to calculate the average number of newton iterations. Say, if you run newmark and want to know the average number of newton raphson iterations per time step. IMO, it doesnt hurt.
